ath79: Add support for D-Link DIR-869

jforman96 opened the pull request:
Add support for dir-869 based on dir-859, shared DTSI like on dir-842-c

Borromini commented:
This has some formal issues, like missing signoff, a clear commit message containing installation instructions and device hardware. For the commit description, please check recent commits like e.g. https://github.com/openwrt/openwrt/commit/1e75909a35a2b361cdfdfcf18a26ad61271b174e.

See also https://openwrt.org/submitting-patches.

I'm not qualified to comment on the actual code, but I do recommend to check what made the old PR die, in case you did not do that yet.

s-2 commented:
Commits should be squashed into individual commits per device, in this case the dtsi split and actual adding of DIR-869 might be separated though.

Things like increasing the spi frequency / updating switch reset gpio for existing devices would usually go into separate commits, though I'm not sure if that's a requirement when the dts would be replaced/renamed later anyways.
However, the commit description should contain a clear indication of these changes.

Code looks good so far, I will start another build later in the evening and test on both devices 🙂

s-2 commented:
Built and tested on DIR-869, works fine so far :slightly_smiling_face: 
LED behaviour is a little awkward with both configured to WAN activity, but this is consistent with ar71xx.
Speaking of which: LEDs were named `:status` in ar71xx but are now `:power`;
the model prefix does not matter here, but either rename these to `:status` as well or add them to
`ath79/generic/base-files/etc/uci-defaults/04_led_migration`
(that's because uci-defaults are only executed on firstboot, but not when sysupgrading from ar71xx to ath79, with settings kept).

Besides, a few more formal issues remain: please check the link provided by @Borromini as well as other devices.
(e.g. dir-869 dts should be moved to the second commit; commit title should start with "ath79: add support..."; maximum length of title / commit message lines; ..)

honza96 commented:
Thank you for the advice, now the problem with the flashing LED should be solved. I also tried to modify the commits, I'll see what else can be added to the information about the router

s-2 commented:
I think the order of LED migration parameters is reversed here, it looks like the old value (`:status`) should belong on the left, the new one (`:power`) to the right side of `=`.

The dts for DIR-869 should go into the second commit, so the first one only contains the split and the frequency / reset-gpio.

Rearranging commits is usually cumbersome, I think the easiest way from here might be to
`git reset --soft HEAD~3`
(beware this will discard your commit messages, so make sure to save them elsewhere first)
`git restore --staged .`
to unstage all changes, so you can edit led migration, finally
`git add <filename>` to pick only the changes for the first commit (dtsi and 859.dts), mention frequency / gpio change in message,
afterwards `git add .` to stage all remaining changes for the second commit.

honza96 commented:
Thanks for the help with the correct procedure, I divided it into 2 commits as you said, today I will try how it works with the led migration fix.

s-2 commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -530,6 +531,7 @@ ath79_setup_macs()
 		label_mac=$lan_mac
 		;;
 	dlink,dir-859-a1|\
+        dlink,dir-869-a1|\
```
indent does not match the rest of the file here (tabs vs. spaces)

s-2 commented:
First commit needs a signoff as well.
Looks like a vim .swp file sneaked into the second commit (I guess that's why people keep saying not to use `git add .` but rather add each file individually :innocent: )

honza96 commented:
Thanks, it should be fine now.

adschm commented:
Looks like the first commit does three different things, so it should be three commits then.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -184,6 +184,10 @@ dlink,dap-1365-a1)
 	;;
+dlink,dir-869-a1)
+	ucidef_set_led_switch "led_power_orange" "WAN" "orange:power" "switch0" "0x20"
+ 	ucidef_set_led_switch "led_power_white" "WAN" "white:power" "switch0" "0x20"
```
Why bind both to the same trigger? Is this vendor behavior?

adschm replied:
Apart from that, if that's a power LED, it should not indicate ethernet at all.

adschm replied:
Ping.

honza96 replied:
I don't know here, is there any list of all triggers?

adschm replied:
If you don't know what you did, let's start with you telling me what you wanted to do ...

honza96 replied:
I was starting from commit # 3296, but I didn't find anything in this file that would be a suitable replacement for "WAN"

honza96 replied:
I've always tried to inspire the dir-859, but there's only one line "green: internet", which is confusing to me because the dir-859 has more LEDs

honza96 replied:
I built as it is here and after the boot, both LEDs blinking. I found that this file affects "/etc/config/system" and accordingly I tried to fix it so that the LEDs work as they should. I renamed it POWER and left only power: white which is the only one used when everything is booted properly, it was also necessary to set the default-on.

adschm replied:
Three comments later, but still I don't know which LEDs the device has and what they are supposed to do.

trsqr replied:
The device has a single LED. Original behavior is described in the manual:
 
![image](https://user-images.githubusercontent.com/6545027/134810067-5a968b2c-8300-4ffe-b9c6-d0516df6a4ed.png)

adschm commented on target/linux/ath79/generic/base-files/etc/uci-defaults/04_led_migration:
```diff
@@ -11,6 +11,10 @@ dlink,dap-1330-a1)
+dlink,dir-869-a1)
+        migrate_leds "orange:status=orange:power" \
+                "white:status=white:power"
+        ;;
```
Why is that?

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -935,6 +935,19 @@ define Device/dlink_dir-859-a1
+  DEVICE_MODEL := DIR-869
+  DEVICE_VARIANT := A1
+  IMAGE_SIZE := 15872k
+  DEVICE_PACKAGES :=  kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct
```
Double space after ":="

adschm commented on target/linux/ath79/image/generic.mk:
```diff
@@ -935,6 +935,19 @@ define Device/dlink_dir-859-a1
+  IMAGE_SIZE := 15872k
+  DEVICE_PACKAGES :=  kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct
+  SEAMA_SIGNATURE := wrgac54_dlink.2015_dir869
+  SUPPORTED_DEVICES += dir-869-a1
```
Why the supported_devices?

honza96 replied:
from the previous PR ( #3296) for DIR-869:

> Please add the line
> ` SUPPORTED_DEVICES += dir-869-a1`
> here, since this device has been supported in ar71xx, but the format of device names was changed for ath79.
> 
> This will add the old model string from ar71xx to the metadata appended to `.sysupgrade` images, allowing routers currently running ar71xx to update smoothly to ath79.
> 
> Just updated mine from LEDE, great to see this device ported to ath79! slightly_smiling_face

Je to tedy v pořádku nebo je nutné to odstranit?

adschm replied:
Keep it, since the referenced statement is true.

s-2 commented:
As mentioned earlier, this device was supported in ar71xx already, thus the LED migration:

> LED behaviour is a little awkward with both configured to WAN activity, but this is consistent with ar71xx.

However personally, I would also prefer to discard the previous LED settings and enforce e.g. white (permanent on) for the status (`led_running`) and orange for the rest, this would also match the vendor behaviour more closely. The user could configure the LED for WAN activity manually, if desired.

adschm commented:
> As mentioned earlier, this device was supported in ar71xx already, thus the LED migration:
> 
> > LED behaviour is a little awkward with both configured to WAN activity, but this is consistent with ar71xx.
> 
> However personally, I would also prefer to discard the previous LED settings and enforce e.g. white (permanent on) for the status (`led_running`) and orange for the rest, this would also match the vendor behaviour more closely. The user could configure the LED for WAN activity manually, if desired.

Since both are used for status, I'd also drop the migration since it's useless with the new binding via led-* in DTS.

Apart from that, what is the reason for renaming status->power?

s-2 commented:
> Apart from that, what is the reason for renaming status->power?
Most probably that DIR-842 / 859 served as a reference for similar ath79 devices here, as well as others predominantly naming it `power`, though I don't have any opinions on this.

honza96 commented:
Hi, I've separated the first confirmation as written here,
The double space after ": =" is also fixed, I also removed led_migration

adschm commented:
Missing commit message for the first three patches. Is the final patch supposed to be squashed?

honza96 commented:
Hi, I added a missing commit message.

adschm commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -134,7 +124,7 @@
 		phy-mode = "sgmii";
 		qca,mib-poll-interval = <500>;
-
+		reset-gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
```
Consider adding empty lines before and after reset-gpios for easier reading.

adschm commented on target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:
```diff
@@ -10,7 +10,7 @@
 		led-boot = &led_power_orange;
 		led-failsafe = &led_power_orange;
 		led-running = &led_power_white;
-		led-upgrade = &led_power_white;
```
Since you add this device, please squash this change into your previous commit.

adschm commented:
Picked the first commit for now.

adschm commented:
> ath79: Replace reset-button, spi-max-frequency increase 

These are two changes, please make them two separate commits.

> Merge branch 'master' into master

Please remove the merge commit (i.e. squash the changes into the proper commit).

adschm commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -40,124 +37,4 @@
-
-	qca,no-eeprom;
-};
+};
```
Missing new line.

adschm commented on target/linux/ath79/generic/base-files/etc/board.d/01_leds:
```diff
@@ -188,6 +188,9 @@ dlink,dap-1365-a1)
 	;;
+dlink,dir-869-a1)
+	ucidef_set_led_switch "led_power_white" "POWER" "white:power" "default-on" "0x20"
+	;;
```
So, now we have power-LEDs instead of internet-LEDs?

honza96 replied:
The LEDs on this router are a little different, it has white and orange LEDs, but they shine from the same place. White LED indicates that the device is booted. I think there is no need to define an orange LED here, because when the device is successfully booted, the orange LED is not used.

honza96 replied:
The name power seems appropriate to me, because in dir-869-a1.dts I named the LEDs power_white and power_orange.

adschm replied:
Power LEDs should be defined in DTS. But apart from that, random config changes without any justification make you look like you don't care about what's correct.

honza96 replied:
I'm sorry if it seemed like a random config change. The previous setting of this file caused both LEDs to flash, I modified /etc/config/system so that the white LED lights up after successful boot and I modified this file accordingly. The red LED is not needed define here because the behaviour in a state other than a successful boot is specified in DTS.

CodeFetch replied:
@honza96 I don't get it, too... So the LEDs are also controlled by AR8327, but you still need this entry although you've defined it in DTS using GPIOs?!

honza96 replied:
hi, i tried build without this entry and LED works fine. Thank you for the advice.

Commit:
ath79: Replace reset-button for DIR-859

gpio-export for the switch reset pin replaced with a reset pin definition for the driver, within the phy node.

Signed-off-by: Jan Forman <forman.jan96@gmail.com>
Tested-By: Sebastian Schaper <openwrt@sebastianschaper.net>

Commit:
ath79: Create shared dtsi for DIR-859

Create a shared dtsi for the dir-859 and similarly device, it similarly as it done for the dir-842.

Signed-off-by: Jan Forman <jforman@tuta.io>

CodeFetch commented on target/linux/ath79/image/generic.mk:
```diff
@@ -935,6 +935,20 @@ define Device/dlink_dir-859-a1
+  DEVICE_PACKAGES := kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct
+  SEAMA_SIGNATURE := wrgac54_dlink.2015_dir869
+  SUPPORTED_DEVICES += dir-869-a1
+This conversation was marked as resolved by honza96
```
I think here is a copy-paste error. And please squash the commits...

adschm commented:
> spi-max-frequency increased to 50000000

Please don't write down the obvious but explain why this can be done (in the commit message).

And please be specific in your commit title, you don't increase _all_ frequencies, but just for a single device.

Same for the subsequent commit. What do you replace with what for what and why ...?

Commit:
ath79: Add support for D-Link DIR-869-A1

Specifications
	The D-Link EXO AC1750 (DIR-869) router released in 2016.
	It is powered by Qualcomm Atheros QCA9563 @ 750 MHz chipset, 64 MB RAM and 16 MB flash.
	10/100/1000 Gigabit Ethernet WAN port
	Four 10/100/1000 Gigabit Ethernet LAN ports
	Power Button, Reset Button, WPS Button, Mode Switch

Flashing
	1. Upload factory.bin via D-link web interface (Management/Upgrade).

Revert to stock
	Upload original firmware via OpenWrt sysupgrade interface.

Debricking
	D-Link Recovery GUI (192.168.0.1)

Signed-off-by: Jan Forman <forman.jan96@gmail.com>

adschm commented:
MAC address setup for ethernet is missing.
Flashing instructions are missing.
Is the address in the commit message really 192.169.0.1 and not 192.168.0.1?

Please group together the specification details in the last commit message into a section "Specifications".

s-2 commented:
```
	Do not turn off the device after starting OpenWrt for the first time.
	Otherwise, u-boot will go into safe mode because the firmware checksum has changed.
	After sysupgrade, u-boot will not complain about the firmware checksum.
```
Is this indeed still required for this device, i.e. could you check flashing to OEM once again and then back to OpenWrt?

I think for DIR-842 and DIR-859 this was no longer required with latest master, otherwise the device might need to be added to
`/base-files/etc/uci-defaults/09_fix-checksum`, though maybe even that only device (Qihoo C301) still using `fix_seama_header` to day might need to be re-checked for flashing issues, if the script were not called.

honza96 commented:
> ```
> 	Do not turn off the device after starting OpenWrt for the first time.
> 	Otherwise, u-boot will go into safe mode because the firmware checksum has changed.
> 	After sysupgrade, u-boot will not complain about the firmware checksum.
> ```
> 
> Is this indeed still required for this device, i.e. could you check flashing to OEM once again and then back to OpenWrt?
> 
> I think for DIR-842 and DIR-859 this was no longer required with latest master, otherwise the device might need to be added to
> `/base-files/etc/uci-defaults/09_fix-checksum`, though maybe even that only device (Qihoo C301) still using `fix_seama_header` to day might need to be re-checked for flashing issues, if the script were not called.


It is necessary on my repository, I will try the last master.

honza96 commented:
Unfortunately, the problem persists with  master.

honza96 commented:
Adding the device to "09_fix-checksum" solves the problem with unsuccessfully starting the device after the first reboot. I also added DIR-859 so it will probably be necessary to change the flash instructions for [DIR-859](https://github.com/openwrt/openwrt/commit/abc7ed2c58860fc1363129efbe9e8c60f81a802c)

s-2 commented:
> I also added DIR-859

I'm not sure if this is still necessary for that device, although I only recall flashing one of these some time ago, and I think there were no issues; however multiple bootloader versions might exist (at least for DIR-842 there are two versions, but IMHO this did not affect image checksum either, but mostly the position of art calibration data).

In any case, I was only wondering whether the script could be removed completely, but since it indeed makes a difference for DIR-869, it still has a reason to be kept (as it may be required for the other device just as well).

adschm commented:
There is still no justification at all for the reset-gpios change ...

s-2 commented:
> There is still no justification at all for the reset-gpios change ...

The boards for DIR-842 Rev. C / DIR-859 / DIR-869 are all based on the same layout and OEM firmware; I had replaced the gpio-hog for DIR-842 Gigabit switch with a proper reset-gpio defintion some time ago (8643c0b53d74aeb535c6700c115d323cb55ec7fb), and I think I recall testing this with DIR-859 as well, however I may as well dig this device back out and re-test with the latest master, so @honza96 could e.g. add my Tested-By to that commit.

Though I'm not quite sure whether the kernel driver will actually use this GPIO at all, to reset a switch in case it is no longer responding, but in any case it seemed to be cleanest way to declare the switch reset pin, when it needs to be pulled high for the switch to work. Many other devices might actually have this pin connected to the SoC as well, but using a pull-up etc. so this is not needed / was never found when adding device support, thus leaving no way to reset a non-responding switch wiuthout power-cycling the device.

adschm commented:
@s-2 Could you just put together a proper patch for that with a meaningful commit message and either link it here or e-mail it to me so I can directly apply it?
Same would be helpful for the SPI frequency change, but that's less dramatic ...

s-2 commented:
> Same would be helpful for the SPI frequency change, but that's less dramatic ...

Weird, this change has actually been in master since April (328546e1088d5b559eddd40a03318bef07f47c92). @honza96 please rebase!

Building latest master with `reset-gpios` now.

s-2 commented:
Tested `reset-gpios` with DIR-859, works with current master. Might as well use the commit from this PR and add my Tested-By etc.

While testing the reset button, it turned out this device still needs either two-stage factory flashing, or having the checksum corrected automatically on factory flashing.
So I would prefer to add this to `09_fix-checksum`. (@honza96 sorry for not testing this earlier, could have left it that way indeed)

https://github.com/s-2/openwrt/commits/dir-859_fixes

Besides, I wonder how many users DIR-859 actually has...

honza96 commented:
"SPI frequency" commit deleted, "reset-gpios" now have Tested-By: @s-2, "09_fix-checksum" commit squashed.

trsqr commented:
What's the status here? My DIR-869 is on 19.07 still... Any help needed for anything?

honza96 commented:
These changes work with version 22.03. @adschm what needs to be done right now? I believe we will be able to fix it.

DragonBluep commented on target/linux/ath79/dts/qca9563_dlink_dir-8x9-a1.dtsi:
```diff
@@ -0,0 +1,121 @@
+&mdio0 {
+	status = "okay";
+
+	phy-mask = <0>;
```
Please drop `phy-mask = <0>;`, we do not need it anymore. https://github.com/openwrt/openwrt/commit/dae392752761128e13ede484217d806d643cbf2f

DragonBluep commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -222,6 +222,7 @@ ath79_setup_interfaces()
 	dlink,dir-842-c2|\
 	dlink,dir-842-c3|\
 	dlink,dir-859-a1|\
+	dlink,dir-869-a1|\
```
And do not forget set mac address in this file. Just like the dir-859.

DragonBluep replied:
Try to use `mtd_get_mac_text` instead of `mtd_get_mac_ascii` so that we can easily set mac address in dts after we convert to DSA driver. `devdata` seems not a writable partition, so I believe mac address offset are fixed.

DragonBluep replied:
Yes, you should separate them. And you also need to find the mac address offset, refer to:
https://github.com/openwrt/openwrt/blob/80a62a675d2ef57a1885e1829926dd2ea1682612/target/linux/ath79/generic/base-files/etc/board.d/02_network#L609
If you feel difficult about it, then you can upload the `devdata` partition here, I can help to analyze it.

honza96 replied:
If I understand it correctly, the offset is 0xc9. I am sending devdata.
[dir-869.mtd2.bin.zip](https://github.com/openwrt/openwrt/files/9863281/dir-869.mtd2.bin.zip)

DragonBluep replied:
> If I understand it correctly, the offset is 0xc9 for lan_mac and 0x79 for wan_mac.

Yes you are right.
Considering that it has been more than one year since you create this PR, would you mind rebase it to current master branch in case we miss some important things?

DragonBluep replied:
I got a flash backup of dir-859 A3. It seems that their mac addresses have different offsets. Unfortunately, we have to revert these changes.
![image](https://user-images.githubusercontent.com/70847398/201042255-badcd9dd-a828-4e64-9a8a-4323313d5955.png)

honza96 replied:
Hi @DragonBluep, can you please send your offset so I can split the devices in the 02_network file?

s-2 replied:
I think this format should use the `mtd_get_mac_ascii` function to parse the equals-separated values, even though this is currently not supported to be set within dts. But technically, the offsets might change even between two devices of the same series (e.g. depending on the exact OEM firmware version that changed).

DragonBluep commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom:
```diff
@@ -32,6 +32,7 @@ case "$FIRMWARE" in
+	dlink,dir-869-a1|\
 	nec,wg1200cr|\
 	wd,mynet-n750)
 		caldata_extract "art" 0x1000 0x440
```
It's a good point to convert calibration data to nvmem, please refer 48625a04453b4f5775db2d2a5abebf977b81344c.
And now we can set ascii mac address in dts, please check if it works. https://github.com/openwrt/openwrt/commit/ecd81de7a5ab4899cc95107b402f7a402caf70bd https://github.com/openwrt/openwrt/commit/5c147d36ba7e998aee7ff4f81553b935bcfc5383 .

DragonBluep replied:
It's worth to rebase upstream/master and try this patch:
<details>
<summary>Convert to nvmem</summary>

```
diff --git a/target/linux/ath79/dts/qca9563_dlink_dir-8x9-a1.dtsi b/target/linux/ath79/dts/qca9563_dlink_dir-8x9-a1.dtsi
index c5fcb0c8b0..36059d81b8 100644
--- a/target/linux/ath79/dts/qca9563_dlink_dir-8x9-a1.dtsi
+++ b/target/linux/ath79/dts/qca9563_dlink_dir-8x9-a1.dtsi
@@ -27,6 +27,14 @@
 
 &pcie {
 	status = "okay";
+
+	wifi@0,0 {
+		compatible = "qcom,ath10k";
+		reg = <0 0 0 0 0>;
+
+		nvmem-cells = <&calibration_ath10k>, <&macaddr_devdata_94>;
+		nvmem-cell-names = "pre-calibration", "mac-address-ascii";
+	};
 };
 
 &spi {
@@ -58,6 +66,18 @@
 				label = "devdata";
 				reg = <0x050000 0x10000>;
 				read-only;
+
+				compatible = "nvmem-cells";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				macaddr_devdata_94: macaddr@94 {
+					reg = <0x94 0x11>;
+				};
+
+				macaddr_devdata_b0: macaddr@b0 {
+					reg = <0xb0 0x11>;
+				};
 			};
 
 			partition@60000 {
@@ -72,10 +92,22 @@
 				reg = <0x070000 0xf80000>;
 			};
 
-			art: partition@ff0000 {
+			partition@ff0000 {
 				label = "art";
 				reg = <0xff0000 0x010000>;
 				read-only;
+
+				compatible = "nvmem-cells";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				calibration_ath9k: calibration@1000 {
+					reg = <0x1000 0x440>;
+				};
+
+				calibration_ath10k: calibration@5000 {
+					reg = <0x5000 0x844>;
+				};
 			};
 		};
 	};
@@ -115,5 +147,6 @@
 &wmac {
 	status = "okay";
 
-	qca,no-eeprom;
+	nvmem-cells = <&calibration_ath9k>, <&macaddr_devdata_b0>;
+	nvmem-cell-names = "calibration", "mac-address-ascii";
 };
diff --git a/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom b/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
index b5f3460652..0746b48290 100644
--- a/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
+++ b/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/10-ath9k-eeprom
@@ -31,8 +31,6 @@ case "$FIRMWARE" in
 	dlink,dir-842-c1|\
 	dlink,dir-842-c2|\
 	dlink,dir-842-c3|\
-	dlink,dir-859-a1|\
-	dlink,dir-869-a1|\
 	nec,wg1200cr|\
 	wd,mynet-n750)
 		caldata_extract "art" 0x1000 0x440
diff --git a/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata b/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
index 3eb3272ab7..ffb62513e1 100644
--- a/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
+++ b/target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata
@@ -61,11 +61,6 @@ case "$FIRMWARE" in
 		caldata_extract "art" 0x5000 0x844
 		ath10k_patch_mac $(mtd_get_mac_ascii bdcfg wlanmac_a)
 		;;
-	dlink,dir-859-a1|\
-	dlink,dir-869-a1)
-		caldata_extract "art" 0x5000 0x844
-		ath10k_patch_mac $(mtd_get_mac_ascii devdata "wlan5mac")
-		;;
 	elecom,wrc-1750ghbk2-i)
 		caldata_extract "art" 0x5000 0x844
 		;;

```
</details>

honza96 replied:
Thanks for the patch, I did the rebase and applied the patch you sent. Unfortunately, after flashing the image, I found out that only 5Ghz WiFi works. I am sending the log and I will try to find where the problem could be.
[kernel.log](https://github.com/openwrt/openwrt/files/9959585/kernel.log)

DragonBluep replied:
Looks like you tested v22.03 branch? `mac-address-ascii` can only work in the current master branch. Maybe you can try to reset it and change the wifi country code.

DragonBluep commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -61,7 +61,8 @@ case "$FIRMWARE" in
+	dlink,dir-859-a1|\
+	dlink,dir-869-a1)
 		caldata_extract "art" 0x5000 0x844
 		ath10k_patch_mac $(mtd_get_mac_ascii devdata "wlan5mac")
```
Same as 10-ath9k-eeprom. Please refer to 8ccbc95d5021a638d6c6611e04dcbec15aae2379.

Commit:
ath79: Convert calibration data to nvmem

For D-link DIR-859 and DIR-869
Replace the mtd-cal-data by an nvmem-cell.
Add the PCIe node for the ath10k radio to the devicetree.
Thanks to DragonBlue for this patch

Signed-off-by: Jan Forman <jforman@tuta.io>

honza96 commented:
After using the master, both wifi adapters worked fine. I tried moving this commit between the "add shared dtsi" and the one that adds the new device, but I don't know if it went as intended.

DragonBluep commented:
@honza96 please rebase it.
```
git remote add upstream https://github.com/openwrt/openwrt.git
git fetch upstream
git rebase upstream/master
git push -f
```

honza96 commented:
Hi @DragonBluep, thanks for the advice, I did the rebase as you requested.

s-2 commented:
There still seem to be a few weird commits after the ones that actually add support for DIR-869, maybe this happened while resolving merge conflicts during rebase.
Try `git rebase -i HEAD~9` and in the opened text editor delete the lines for the commits
```
515566d ath79: add support for NEC Aterm WF1200CR
736c659 ath79: add support for WD My Net N600
fa476c4 ath79: drop phy-mask property
```
first, as these would just revert changes that were made tomaster in the meantime.

`4ba7281` removes an artifact from a merge conflict introduced earlier, it should vanish automatically when fixing
`39a7f2f ath79: increase max SPI clock for DIR-859 A1`
I'm not quite sure how to salvage that one though, this should only touch a single line changing the spi frequency, so this should most probably happen to the existing device, before splitting it into a common .dtsi for both devices.
I'm not sure what would happen by just moving the commit to the top during interactive rebase (better have a backup branch before trying anything), it might even be easier to recreate the commit by splitting the .dtsi one, but then the timestamp would be lost.

honza96 commented:
It should be fine now, thanks.

s-2 commented:
DIR-869 looks good now, however the "shared dtsi" commit only adds the dtsi that is used for DIR-869 later, but does not actually remove the redundant contents from the DIR-859 dts that are now available from the dtsi? Also it seems to re-introduce the reset-button declaration that was replaced by `reset-gpios` in the previous commit.

honza96 commented:
I probably did something wrong with the conflict resolution, it's fixed now.

s-2 commented:
Almost :slightly_smiling_face: the 859 dts should (only) `#include "qca9563_dlink_dir-8x9-a1.dtsi"`, also needs a closing `};`.
Rest looks good to me so far.

hauke commented on target/linux/ath79/generic/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:
```diff
@@ -69,9 +69,14 @@ case "$FIRMWARE" in
+		caldata_extract "art" 0x5000 0x844
+		ath10k_patch_mac $(mtd_get_mac_ascii bdcfg wlanmac_a)
+		;;
+	elecom,wrc-1750ghbk2-i)
```
Why do you add other devices in this pull request?

hauke commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -41,112 +41,3 @@
 		};
 	};
 
-	keys {
```
Please  add `#include "qca9563_dlink_dir-8x9-a1.dtsi"` into this file too. 

Please also compile test all the boards you modified.

s-2 commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-or-later OR MIT
 
 #include "qca956x.dtsi"
```
This is already included by `qca9563_dlink_dir-8x9-a1.dtsi`.

honza96 commented:
Thanks for your advice, I've fixed all the typos you pointed out.

hauke commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -1,9 +1,4 @@
-// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```
Please keep the license header

hauke commented on target/linux/ath79/dts/qca9563_dlink_dir-859-a1.dts:
```diff
@@ -1,9 +1,4 @@
-
-#include "qca956x.dtsi"
-
-#include <dt-bindings/gpio/gpio.h>
```
GPIOs are still used in this file, please keep `#include <dt-bindings/gpio/gpio.h>`

s-2 replied:
Is this indeed required here, or more of a stylistic issue?
I see many devices only including these in the dtsi.

hauke replied:
It is stylistic. We will also get it over the include, but I think it is better to include the headers for the stuff used directly.

hauke commented on target/linux/ath79/dts/qca9563_dlink_dir-8x9-a1.dtsi:
```diff
@@ -0,0 +1,119 @@
+			debounce-interval = <60>;
+		};
+	};
+
```
Please remove this empty line

hauke commented on target/linux/ath79/dts/qca9563_dlink_dir-8x9-a1.dtsi:
```diff
@@ -0,0 +1,119 @@
+		reg = <0>;
+		phy-mode = "sgmii";
+		qca,mib-poll-interval = <500>;
+		
```
Please remove trailing tabs.

hauke commented on target/linux/ath79/dts/qca9563_dlink_dir-8x9-a1.dtsi:
```diff
@@ -0,0 +1,119 @@
+		qca,mib-poll-interval = <500>;
+		
+		reset-gpios = <&gpio 11 GPIO_ACTIVE_LOW>;
+		
```
Please remove trailing tabs.

hauke commented on target/linux/ath79/image/generic.mk:
```diff
@@ -1115,6 +1115,19 @@ define Device/dlink_dir-859-a1
+  DEVICE_MODEL := DIR-869
+  DEVICE_VARIANT := A1
+  IMAGE_SIZE := 15872k
+  DEVICE_PACKAGES := kmod-ath10k-ct-smallbuffers ath10k-firmware-qca988x-ct
```
If this device has USB, please add `kmod-usb2` too.

hauke commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -639,6 +640,10 @@ ath79_setup_macs()
+	dlink,dir-869-a1)
+                lan_mac=$(mtd_get_mac_text "devdata" 0xc9)
+                wan_mac=$(mtd_get_mac_text "devdata" 0x79)
+		;;
```
Why are these devices use a different way to extract the mac addresses now?

honza96 replied:
Hi, can you tell me more about what you mean please?

hauke replied:
This change also affects the `dlink,dir-859-a1` device. If the `dlink,dir-859-a1` is detected it will also go into your new condition. If this is not done intentional please change the code like this:
````
 		wan_mac=$(mtd_get_mac_ascii devdata "wanmac")
 		label_mac=$lan_mac
 		;;
+	dlink,dir-869-a1)
+		lan_mac=$(mtd_get_mac_text "devdata" 0xc9)
+		wan_mac=$(mtd_get_mac_text "devdata" 0x79)
+		;;
 	dlink,dir-859-a1|\
 	qihoo,c301|\
 	wd,mynet-n600|\
````

Please also use tabs for indention and not spaces.

hauke replied:
This is still open

honza96 replied:
Thanks for the help, it should be fine now.

hauke commented on target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:
```diff
@@ -0,0 +1,31 @@
+			label = "white:power";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+
```
Please remove empty line

hauke commented on target/linux/ath79/dts/qca9563_dlink_dir-869-a1.dts:
```diff
@@ -0,0 +1,31 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qca9563_dlink_dir-8x9-a1.dtsi"
+
```
Please add `#include <dt-bindings/gpio/gpio.h>`, this file uses some GPIO contants

hauke commented:
Please rebase on top of current master and fix the open comment.

honza96 commented:
Hello @hauke, what else needs to be done?

trsqr commented:
Are there some blocking points still? Can I do something to help? My Dir-869 is still on 19.07 and needs desperately an update...

s-2 reviewed (commented):
The code and order of commits look good, though I'm still not sure about using hardcoded `devdata` offsets for MAC addresses (c.f. https://github.com/openwrt/openwrt/pull/3997/files#r1071632465). If this is only an issue with DIR-859 A3 as mentioned by @DragonBluep, we might just add a new profile for that device, but in general it would feel safer to use the existing parser, even though this can not be used from within the device tree.

I could also flash my device to a current build, though from looking at the code I really don't see any more issues.

honza96 commented:
@s-2  So I set mtd_get_mac_ascii for dir-869 as well.
I flash the current build and everything looks fine

s-2 commented:
So we are setting LAN / WAN MAC the traditional way using the parser, but wifi remains with the fixed offsets.
@DragonBluep did you only notice a difference in offsets for the physical ports of the A3 revision, or did this affect wifi as well? I don't really want to make a decision which way to use here (though I hope one day nvmem might be able to perform these kinds of parsing in-kernel).

DragonBluep commented:
> @DragonBluep did you only notice a difference in offsets for the physical ports of the A3 revision

Yes, only mac offsets.

> I don't really want to make a decision which way to use here (though I hope one day nvmem might be able to perform these kinds of parsing in-kernel).

Maybe the traditional way is better since we don't know the mac offsets of the DIR-859 **A1**.

s-2 commented:
> Yes, only mac offsets.

Just WAN / LAN, or also Wifi MACs? Maybe those should use the traditional way as well, in the current state we have a mix of both.

honza96 commented:
If the better option is "mtd_get_mac_text", DIR-859 A1 and DIR-869 A1 are the same, isn't it better to create a new profile for DIR-859 A3? Or is there a possibility that 2 identical A1 devices will have a different offset?

DragonBluep commented:
> > Yes, only mac offsets.
> 
> Just WAN / LAN, or also Wifi MACs? Maybe those should use the traditional way as well, in the current state we have a mix of both.

All MAC addresses have different offset.
https://github.com/openwrt/openwrt/pull/3997#discussion_r1018796917
macaddr:
wlan2g: devdata@0x5b
wlan5g: devdata@0x76
lan: devdata@0x3f
wan: devdata@0x8f
I have verified these offsets on two different DIR-859 A3.

> Or is there a possibility that 2 identical A1 devices will have a different offset?

This is also what I am worried about. 

> isn't it better to create a new profile for DIR-859 A3?

Yes, if you like, you can add it here by the way. I can help test it.

DragonBluep commented on target/linux/ath79/generic/base-files/etc/board.d/02_network:
```diff
@@ -655,7 +657,15 @@ ath79_setup_macs()
+		wan_mac=$(mtd_get_mac_text "caldata" 0x8f)
+		;;
 	dlink,dir-859-a1|\
+	dlink,dir-869-a1|)
```
Syntax error here.

DragonBluep commented:
I found a better way to configure the MAC address of DIR-859 A3, so I opened a new PR https://github.com/openwrt/openwrt/pull/12534.

honza96 commented:
I removed the commit for DIR-859 A3 and fixed the typo. Is there anything else that needs to be done?
